File: frontend/src/types/emotion.ts
// --- Emotion Vector Analysis ---

export interface EmotionActivation {
  emotion: string;
  activation: number;
  activation_normalized: number;
}

export interface ExtractProbesResponse {
  model_id: string;
  layer_idx: number;
  mode: string;
  n_emotions: number;
  emotions: string[];
  metadata: Record<string, unknown>;
}

export interface SteerComparison {
  original_text: string;
  steered_text: string;
  emotion: string;
  strength: number;
  layer_range: number[];
}

export interface SteerResponse {
  model_id: string;
  prompt: string;
  comparison: SteerComparison;
  original_emotions: EmotionActivation[];
  steered_emotions: EmotionActivation[];
  metadata: Record<string, unknown>;
}

// --- Token-level Projection ---

export interface TokenEmotionProfile {
  token_idx: number;
  token_str: string;
  activations: Record<string, number>;
}

export interface DisplayToken {
  label: string;
  token_indices: number[];
  activations: Record<string, number>;
}

export interface ProjectResponse {
  model_id: string;
  prompt: string;
  layer_idx: number;
  emotions: string[];
  tokens: TokenEmotionProfile[];
  display_tokens: DisplayToken[];
  metadata: Record<string, unknown>;
}

// --- PCA ---

export interface PCAResponse {
  model_id: string;
  layer_idx: number;
  emotions: string[];
  pc1: number[];
  pc2: number[];
  variance_explained: number[];
  metadata: Record<string, unknown>;
}

// --- Sweep ---

export interface SweepPoint {
  strength: number;
  target_emotion_activation: number;
  generated_text: string;
}

export interface SweepResponse {
  model_id: string;
  prompt: string;
  emotion: string;
  points: SweepPoint[];
  metadata: Record<string, unknown>;
}

// --- Layer Evolution ---

export interface LayerEmotionPoint {
  layer_idx: number;
  activations: Record<string, number>;
}

export interface LayerEvolutionResponse {
  model_id: string;
  prompt: string;
  token_idx: number;
  token_str: string;
  emotions: string[];
  layers: LayerEmotionPoint[];
  metadata: Record<string, unknown>;
}

// --- Steer + SAE ---

export interface SAEFeatureDiff {
  feature_idx: number;
  original_activation: number;
  steered_activation: number;
  diff: number;
}

export interface SteerSAEResponse {
  model_id: string;
  prompt: string;
  emotion: string;
  strength: number;
  sae_layer_idx: number;
  sae_hook_name: string;
  d_sae: number;
  original_top_features: SAEFeatureDiff[];
  steered_top_features: SAEFeatureDiff[];
  top_changed_features: SAEFeatureDiff[];
  metadata: Record<string, unknown>;
}

// --- List ---

export interface EmotionListResponse {
  emotions: string[];
  n_emotions: number;
  passages_per_emotion: Record<string, number>;
  model_id: string | null;
  has_probes: boolean;
  probe_layers: number[];
}

File: backend/neural_mri/schemas/emotion.py
from __future__ import annotations

from pydantic import BaseModel, Field

# --- Probe Extraction ---


class ExtractProbesRequest(BaseModel):
    layer_idx: int | None = None  # None = auto (n_layers * 2 // 3)
    mode: str = "comprehension"  # "comprehension" (base+instruct) or "generation" (instruct only)


class EmotionProbe(BaseModel):
    emotion: str
    layer_idx: int
    n_passages: int  # how many passages were averaged


class ExtractProbesResponse(BaseModel):
    model_id: str
    layer_idx: int
    mode: str
    n_emotions: int
    emotions: list[str]
    metadata: dict


# --- Steering ---


class SteerRequest(BaseModel):
    prompt: str
    emotion: str  # which emotion vector to steer with
    strength: float = Field(default=0.05, ge=-1.0, le=1.0)  # fraction of residual stream norm
    layer_range: list[int] | None = None  # None = all layers
    max_new_tokens: int = Field(default=50, ge=1, le=200)


class SteerComparison(BaseModel):
    """Side-by-side: original vs steered output."""

    original_text: str
    steered_text: str
    emotion: str
    strength: float
    layer_range: list[int]


class EmotionActivation(BaseModel):
    emotion: str
    activation: float  # dot product with emotion vector
    activation_normalized: float  # 0-1


class SteerResponse(BaseModel):
    model_id: str
    prompt: str
    comparison: SteerComparison
    # Emotion probe activations on steered vs original (at last token)
    original_emotions: list[EmotionActivation]
    steered_emotions: list[EmotionActivation]
    metadata: dict


# --- Steer + SAE ---


class SteerSAERequest(BaseModel):
    prompt: str
    emotion: str
    strength: float = Field(default=0.05, ge=-1.0, le=1.0)
    sae_layer_idx: int | None = None  # None = auto (probe layer)
    top_k: int = Field(default=20, ge=1, le=100)
    layer_range: list[int] | None = None


class SAEFeatureDiff(BaseModel):
    feature_idx: int
    original_activation: float
    steered_activation: float
    diff: float  # steered - original


class SteerSAEResponse(BaseModel):
    model_id: str
    prompt: str
    emotion: str
    strength: float
    sae_layer_idx: int
    sae_hook_name: str
    d_sae: int
    # Per-token original vs steered top features (at last prompt token)
    original_top_features: list[SAEFeatureDiff]
    steered_top_features: list[SAEFeatureDiff]
    # Features with largest activation change
    top_changed_features: list[SAEFeatureDiff]
    metadata: dict
